pkg/updater/download.go

```diff
@@ -0,0 +1,127 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package updater
+
+import (
+	"archive/tar"
+	"bytes"
+	"compress/gzip"
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+)
+
+const (
+	agentArchiveFileName = "agent.tar.gz"
+)
+
+// downloader is the downloader used by the updater to download packages.
+type downloader struct {
+	client *http.Client
+}
+
+// newDownloader returns a new Downloader.
+func newDownloader(client *http.Client) *downloader {
+	return &downloader{
+		client: client,
+	}
+}
+
+// Download downloads the package at the given URL in temporary directory,
+// verifies its SHA256 hash and extracts it to the given destination path.
+// It currently assumes the package is a tar.gz archive.
+func (d *downloader) Download(ctx context.Context, pkg Package, destinationPath string) error {
+	tmpDir, err := os.MkdirTemp("", "")
+	if err != nil {
+		return fmt.Errorf("could not create temporary directory: %w", err)
```

Comment:
I followed the usual pattern of logging what happens locally so this would be redundant as the above layer will also explain it failed downloading:
https://github.com/DataDog/datadog-agent/blob/46d8fd4266ad5e95839ae90230e7a4b2f4ee68a6/pkg/updater/updater.go#L94